At 4880 popover has incorrect orientation

mzbeetnoff opened the pull request:
Fix issue with Popover component.

What
The popover control previously was not detecting collisions properly. Correct the collision detection, and ensure we call closest scrollable on the target element to avoid clipping popover with edge of container.

Rename `parent` to `container` as it makes more sense with this naming.

Add new menu story to demo passing the `parentRef` to the popover control. Elements in Avail have not utilized this property yet, this storybook now demonstrates the correct usage by setting the `ref` on the target element. As we start to make Avail more accessible this will be very useful to ensure proper keyboard navigation.

Add unit tests

https://user-images.githubusercontent.com/106633502/189464994-9ea947f1-8918-4ea5-af0c-35ff6bccaf4f.mov

quaardvark reviewed (commented):
Code looks good, but don't have bandwith to test thorougly right now

pkrawc commented:
Will pull this down and review today!

pkrawc commented on packages/popover/src/index.js:
```diff
@@ -10,36 +10,42 @@ import {
+}) {
+  const defaultValue = { top: 0, left: 0, visibility: "hidden" }
+  if (!popover || !target || !container) return defaultValue
+  const paddingX = 12
+  const paddingY = 12
```
Should these be user editable? A prop on the component perhaps with default values?

mzbeetnoff replied:
I'm not sure it's necessary? 🤔
A consumer of this component can style the padding and margins of the target element which should have the same effect. Or am I'm missing the idea completely, what are you thinking?

pkrawc replied:
My question then would be why even calculate for it here? It's nice to have defaults but if it can be left up to the consumer then why make it part of the position calc?

mzbeetnoff replied:
I think I see what you're saying and it makes sense. I will make these user editable with default values in the next commit.

mzbeetnoff replied:
Added the configurable property `paddingY` to the `Popover` component. I didn't add a `paddingX` property as I can't see how this would be used currently. The popover always opens either above or below the target, but is always aligned to the left or right side of the menu target.

pkrawc reviewed (changes requested):
I'm not sure if this existed before or not but if there is no position relative on the body--assuming that's the container--then the position on the popover is calculated incorrectly.

`closestScrollable` should return the body if nothing lower is scrollable, but something weird (read I don't understand what) is happening with the box-model when no relative positioning is defined.

mzbeetnoff commented:
> I'm not sure if this existed before or not but if there is no position relative on the body--assuming that's the container--then the position on the popover is calculated incorrectly.
> 
> `closestScrollable` should return the body if nothing lower is scrollable, but something weird (read I don't understand what) is happening with the box-model when no relative positioning is defined.

@pkrawc I believe I have corrected the issue. As far as I can tell the issue has existed for sometime, I was able to verify this by checking out an older `Popover` version  [here](https://github.com/rentalutions/elements/blob/697d0bec3744266baaf8cd0d430690e0dd45be7d/packages/popover/src/index.js) and using the example story you included to demonstrate it. It seems the issue is related to setting the margin value to something other than 0, and how the resize handler calculates the targetBounds.

![image](https://user-images.githubusercontent.com/106633502/192386408-3d2510d5-d386-49d7-80a8-193efb013d4e.png)

mzbeetnoff commented:
> I'm not sure if this existed before or not but if there is no position relative on the body--assuming that's the container--then the position on the popover is calculated incorrectly.
> 
> `closestScrollable` should return the body if nothing lower is scrollable, but something weird (read I don't understand what) is happening with the box-model when no relative positioning is defined.

@pkrawc I believe I have corrected the issue. As far as I can tell the issue has existed for sometime, I was able to verify this by checking out an older `Popover` version  [here](https://github.com/rentalutions/elements/blob/697d0bec3744266baaf8cd0d430690e0dd45be7d/packages/popover/src/index.js) and using the example story you included to demonstrate it. It seems the issue is related to setting the margin value to something other than 0, and how the resize handler calculates the targetBounds.

![image](https://user-images.githubusercontent.com/106633502/192386408-3d2510d5-d386-49d7-80a8-193efb013d4e.png)

mzbeetnoff commented on packages/popover/src/index.js:
```diff
@@ -80,32 +124,43 @@ const Popover = forwardRef(function Popover(
+  const defaultPosition = { top: 0, left: 0, visibility: "hidden" }
+  const portalTarget = usePortal(undefined, parentRef?.current || container)
+  const popoverBounds = useResize(popoverRef, container)
+  const targetBounds = useResize(targetRef, targetRef.parent)
```
This is how I was able to fix the issue mentioned in the comments. Instead of passing the `container` to the resize handler, I instead now pass the targetRef parent.

mzbeetnoff commented:
> > I'm not sure if this existed before or not but if there is no position relative on the body--assuming that's the container--then the position on the popover is calculated incorrectly.
> > `closestScrollable` should return the body if nothing lower is scrollable, but something weird (read I don't understand what) is happening with the box-model when no relative positioning is defined.
> 
> @pkrawc I believe I have corrected the issue. As far as I can tell the issue has existed for sometime, I was able to verify this by checking out an older `Popover` version [here](https://github.com/rentalutions/elements/blob/697d0bec3744266baaf8cd0d430690e0dd45be7d/packages/popover/src/index.js) and using the example story you included to demonstrate it. It seems the issue is related to setting the margin value to something other than 0, and how the resize handler calculates the targetBounds.
> 
> ![image](https://user-images.githubusercontent.com/106633502/192386408-3d2510d5-d386-49d7-80a8-193efb013d4e.png)

@pkrawc I reverted the change that fixed your popover example story because it was breaking the menu portal usage story and also because I feel the fix should be made in useResize. I'm kind of thinking we should revisit this issue separately in a later pull request / jira because it seems out of scope with this change. what do you think?

pkrawc reviewed (approved):
Cool. As long as we're thinking it won't effect any consumers (by introducing something new that they don't expect) I agree that it's likely outside the scope of this fix. Thank you for following up on all this.
